clients/cobol-lsp-vscode-extension/src/services/copybook/downloader/ZoweExplorerDownloader.ts

```diff
@@ -0,0 +1,170 @@
+
+export abstract class ZoweExplorerDownloader {
+  protected explorerAPI: IApiRegisterClient | undefined;
+  private static lockedProfile: Set<string> = new Set();
+
+  constructor(private readonly storagePath: string) {
+    this.loadZEApis();
+  }
+
+  private async loadZEApis() {
+    this.explorerAPI = await Utils.getZoweExplorerAPI();
+  }
+
+  protected abstract downloadCopybookContent(
+    dataset: string,
+    member: string,
+    profileName: string,
+  ): Promise<void>;
+
+  protected getZoweProfile(profileName: string) {
+    return this.explorerAPI!.getExplorerExtenderApi()
+      .getProfilesCache()
+      .loadNamedProfile(profileName);
+  }
+
+  protected getDownloadOptions(
+    profileName: string,
+    dataset: string,
+    member: string,
+    loadedProfile: IProfileLoaded,
+  ) {
+    const downloadBinary = !!SettingsService.getCopybookFileEncoding();
+    const baseUri = vscode.Uri.file(
+      CopybookURI.createDatasetPath(profileName, dataset, this.storagePath),
+    );
+    const filePath = vscode.Uri.joinPath(
+      baseUri,
+      DownloadUtil.getFilenameWithoutExtension(member),
+    );
+
+    const downloadOptions = {
+      file: filePath.fsPath,
+      binary: downloadBinary,
+      returnEtag: true,
+    };
+
+    if (!SettingsService.getCopybookFileEncoding()) {
+      (downloadOptions as any).encoding = loadedProfile.profile.encoding;
+    }
+
+    return { downloadOptions, filePath, downloadBinary };
+  }
+
+  protected async encodeDownloadedContent(
+    filePath: vscode.Uri,
+    downloadBinary: boolean,
+    insertNewLineOn80Char: boolean = false,
+  ) {
+    const fileContents = await vscode.workspace.fs.readFile(filePath);
+    if (downloadBinary) {
+      let newContent = iconv.decode(
+        Buffer.from(fileContents),
+        SettingsService.getCopybookFileEncoding() as string,
+      );
+
+      if (insertNewLineOn80Char) {
+        // Based on assumption - Most of source code on z/OS is 80 characters per record - JCL, HLASM, COBOL
+        // Can be exposed later on as a setting.
+        newContent = newContent.replace(/.{80}/g, `$&\n`);
+      }
+      const writeData = Buffer.from(newContent, "utf8");
+      await vscode.workspace.fs.writeFile(filePath, writeData);
+    }
+  }
+
+  protected async downloadCopybookFromMFUsingZowe(
+    dataset: string,
+    member: string,
+    profileName: string,
+  ): Promise<boolean> {
+    const copybookPath = CopybookURI.createCopybookPath(
+      profileName,
+      dataset,
+      member,
+      this.storagePath,
+    );
+
+    if (!(await DownloadUtil.checkPathExists(copybookPath))) {
+      try {
+        await this.downloadCopybookContent(dataset, member, profileName);
+        return true;
+      } catch (err: any) {
+        vscode.window.showErrorMessage(err.message);
+      }
+    }
+
+    return false;
+  }
+
+  protected handleError(e: any, profileName: string) {
+    if (this.isInvalidCredentials(e)) {
```

Comment:
Given the async nature of the caller, I don't think this really solves the account lock-out in the scenario of many concurrent requests like after re-opening a workspace.

Reply:
I still need to figure this out. will change this in next commit

Reply:
updated at cd9304a335d80598f79983af336e3dac4f1e4d88
Concurrency level is reduced but the issue remains